content/2016.12.how-to-develop-a-webapp-with-Kinto.md

```diff
@@ -0,0 +1,235 @@
+{.image}
+
+* A really simple chat webapp with minimal UI.
+* We will need a bit of JavaScript to enhance and animate interactions within the application and we will also need a way to store and share the app's data. Our front-end should be able to consume an API that would look like `https://api.discute.re/v1/rooms/4815162342/messages` with [discute.re](http://discute.re) as the domain of our application (*discutere* means to chat in Latin).
+
+Our API endpoints could be:
+
+* `/rooms` to return all chat rooms,
+* `/rooms/:id/` to return a particular chat room identified by an ID,
+* `/rooms/:id/messages` to return all messages from a particular room.
+
+To post a message, our front-end will do a POST request to `/rooms/:id/messages`.
+
+We will store each message like this:
+
+    {
+        "message": "I know nothing",
+        "last_modified": 1480340246147,
+        "author": "Jon Snow",
+        "id": "a1e6e879-5c94-4a01-9ec2-28e152d1d8f4"
+    }
+
+
+The JavaScript code could use whatever library or latest hype JS framework but in our example we will stick with modern [Vanilla JS](http://vanilla-js.com/) and use top notch [ES2015 syntax](https://babeljs.io/docs/learn-es2015/).
+
+## API creation
+
+Hmm… what backend should we use? Django? Express? Firebase?… oh wait, why don’t we give [Kinto](http://www.kinto-storage.org/) a try?
+
+Kinto is a perfect fit for this application as once installed, Kinto provides a flexible enough API tohandle other APIs.
+
+Let's see how to use Kinto for our instant chat WebApp.
+
+To show some pride, we will use an explicit subdomain for our main URL `https://kinto.discute.re/v1` to show our configuration rather than an agnostic https://api.discute.re/v1.
+
+Here are the different routes:
+
+- `/rooms/4815162342` becomes `/collections/4815162342` ;
+- `/rooms/4815162342/messages` becomes `/collections/4815162342/records` ;
+
+As we have seen before, we can add a new post to our backend with a simple POST request
+`/collections/4815162342/records`.
+
+A Kinto server can handle many Web applications through the concept of [*buckets*](https://kinto.readthedocs.io/en/latest/concepts.html#buckets-collections-and-records); a bucket can contain many *collections*, and a collection can contain many *records*, and records can contain arbitrary JSON data.
+
+![](https://kinto.readthedocs.io/en/latest/_images/concepts-general.png)
+{.image}
+
+For our application, we’ll create a `chat` bucket, which will contain as many collections as we want distinct chat rooms. Each room collection will itself represent chat messages through records.
+
+You can use [kinto-admin](http://kinto.github.io/kinto-admin/) plugin to easily manage your buckets, collections and records.
+
+![](https://d2mxuefqeaa7sj.cloudfront.net/s_4340C716162DF03B2B97575BFA438C542185694D96D86DB6EB44B989957A3ADE_1480537208814_file.png)
+{.image}
+
+So the request that will allow retrieving all messages from our example  `4815162342` room is:
```

Comment:
With this in mind:

- We'll retrieve all messages for a `4815162342` room by requesting `GET https://kinto.discute.re/v1/buckets/chat/collections/4815162342/records`
- We'll post new messages to the room using `POST https://kinto.discute.re/buckets/chat/collections/4815162342/records`